base_layer/mmr/src/merkle_proof.rs

```diff
@@ -0,0 +1,275 @@
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Portions of this file were originally copyrighted (c) 2018 The Grin Developers, issued under the Apache License,
+// Version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0.
+
+use crate::{
+    backend::ArrayLike,
+    common::{family, family_branch, find_peaks, hash_together, is_leaf, is_left_sibling, leaf_index},
+    Hash,
+    HashSlice,
+    MerkleMountainRange,
+};
+use derive_error::Error;
+use digest::Digest;
+use log::error;
+use serde::{Deserialize, Serialize};
+use std::fmt::{self, Display, Formatter};
+use tari_utilities::hex::{to_hex, Hex};
+
+/// Merkle proof errors.
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum MerkleProofError {
+    // Merkle proof root hash does not match when attempting to verify.
+    RootMismatch,
+    // You tried to construct or verify a Merkle proof using a non-leaf node as the inclusion candidate
+    NonLeafNode,
+    // There was no hash in the merkle tree backend with the given position
+    #[error(non_std, no_from)]
+    HashNotFound(usize),
+    // The list of peak hashes provided in the proof has an error
+    IncorrectPeakMap,
+    // Unexpected
+    Unexpected,
+}
+
+/// A Merkle proof that proves a particular element at a particular position exists in an MMR.
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone, PartialOrd, Ord)]
+pub struct MerkleProof {
+    /// The size of the MMR at the time the proof was created.
+    mmr_size: usize,
+    /// The sibling path from the leaf up to the final sibling hashing to the local root.
+    path: Vec<Hash>,
+    /// The set of MMR peaks, not including the local peak for the candidate node
+    peaks: Vec<Hash>,
```

Comment:
do we need to store the peaks?
Because for a set size there will always be a set amount of peaks, they have a direct relationship.

Reply:
These are the hashes of the peaks that form part of the Merkle Proof